File: app/src/collections/fizzbuzz-list.ts
import { FizzBuzzValue } from '../value-objects/fizzbuzz-value';
import { FizzBuzzType } from '../types/fizzbuzz-type';
import { IndexOutOfRangeError } from '../exceptions/custom-errors';

// ファーストクラスコレクション
export class FizzBuzzList {
  private readonly values: readonly FizzBuzzValue[];

  constructor(values: FizzBuzzValue[] = []) {
    // 防御的コピーによる不変性確保
    this.values = Object.freeze([...values]);
  }

  // 要素の追加（新しいインスタンスを返す）
  add(value: FizzBuzzValue): FizzBuzzList {
    return new FizzBuzzList([...this.values, value]);
  }

  // 複数要素の追加
  addAll(values: FizzBuzzValue[]): FizzBuzzList {
    return new FizzBuzzList([...this.values, ...values]);
  }

  // 要素の取得
  get(index: number): FizzBuzzValue {
    if (index < 0 || index >= this.values.length) {
      throw new IndexOutOfRangeError(index, this.values.length);
    }
    return this.values[index];
  }

  // 安全な要素取得（Optional的な動作）
  tryGet(index: number): FizzBuzzValue | undefined {
    if (index < 0 || index >= this.values.length) {
      return undefined;
    }
    return this.values[index];
  }

  // 長さ
  get length(): number {
    return this.values.length;
  }

  // 文字列配列として取得（後方互換性のため）
  toStringArray(): string[] {
    return this.values.map(v => v.value);
  }

  // iterator
  *[Symbol.iterator](): Generator<FizzBuzzValue, void, unknown> {
    for (const value of this.values) {
      yield value;
    }
  }

  // 範囲で作成（ファクトリメソッド）
  static createRange(fizzBuzzType: FizzBuzzType, start: number = 1, end: number = 100): FizzBuzzList {
    const values: FizzBuzzValue[] = [];
    for (let i = start; i <= end; i++) {
      values.push(fizzBuzzType.generate(i));
    }
    return new FizzBuzzList(values);
  }

  // 統計情報を取得
  getStatistics(): { fizz: number; buzz: number; fizzBuzz: number; numbers: number } {
    const stats = { fizz: 0, buzz: 0, fizzBuzz: 0, numbers: 0 };
    for (const value of this.values) {
      if (value.isFizzBuzz()) stats.fizzBuzz++;
      else if (value.isFizz()) stats.fizz++;
      else if (value.isBuzz()) stats.buzz++;
      else stats.numbers++;
    }
    return stats;
  }

  // フィルタリング
  filter(predicate: (value: FizzBuzzValue) => boolean): FizzBuzzList {
    return new FizzBuzzList(this.values.filter(predicate));
  }

  // FizzBuzzのみを取得
  onlyFizzBuzz(): FizzBuzzList {
    return this.filter(value => value.isFizzBuzz());
  }

  // 文字列表現
  toString(): string {
    return this.values.map(v => v.toString()).join(', ');
  }

  // JSON表現
  toJSON(): string[] {
    return this.values.map(v => v.toJSON());
  }

  // 等価性
  equals(other: FizzBuzzList): boolean {
    if (!(other instanceof FizzBuzzList)) {
      return false;
    }
    if (this.length !== other.length) {
      return false;
    }
    for (let i = 0; i < this.length; i++) {
      if (!this.get(i).equals(other.get(i))) {
        return false;
      }
    }
    return true;
  }
}

File: app/fizzbuzz-scala/src/main/scala/domain/model/FizzBuzzList.scala
package domain.model

// 純粋な不変データ構造として定義
case class FizzBuzzList private (values: List[FizzBuzzValue]) {
  // 不変な操作のみ提供
  def add(value: FizzBuzzValue): FizzBuzzList = 
    FizzBuzzList.create(values :+ value)
  
  def get(index: Int): Option[FizzBuzzValue] = 
    values.lift(index)
  
  def toArray: Array[String] = 
    values.map(_.value).toArray
  
  def map[B](f: FizzBuzzValue => B): List[B] = 
    values.map(f)
  
  def filter(predicate: FizzBuzzValue => Boolean): FizzBuzzList = 
    FizzBuzzList.create(values.filter(predicate))
  
  def foldLeft[B](z: B)(op: (B, FizzBuzzValue) => B): B = 
    values.foldLeft(z)(op)
  
  def size: Int = values.length
  
  override def toString: String = 
    values.map(_.toString).mkString(",")
}

object FizzBuzzList {
  val MaxCount = 100
  
  // スマートコンストラクタで検証
  def create(values: List[FizzBuzzValue]): FizzBuzzList = {
    require(values.length <= MaxCount, s"上限は${MaxCount}件までです")
    new FizzBuzzList(values)
  }
  
  // 関数型的なファクトリメソッド
  def generate(start: Int, end: Int, fizzBuzzType: domain.types.FizzBuzzType): FizzBuzzList = {
    val values = (start to end).map(fizzBuzzType.generate).toList
    create(values)
  }
  
  // 空のリスト
  def empty: FizzBuzzList = 
    new FizzBuzzList(List.empty)
  
  // 単一要素のリスト
  def single(value: FizzBuzzValue): FizzBuzzList = 
    create(List(value))
  
  // 関数合成を使用したビルダー
  def builder: FizzBuzzListBuilder = 
    new FizzBuzzListBuilder(List.empty)
}

// 関数型ビルダーパターン
class FizzBuzzListBuilder private[model] (private val values: List[FizzBuzzValue]) {
  def add(value: FizzBuzzValue): FizzBuzzListBuilder = 
    new FizzBuzzListBuilder(values :+ value)
  
  def addAll(newValues: List[FizzBuzzValue]): FizzBuzzListBuilder = 
    new FizzBuzzListBuilder(values ++ newValues)
  
  def build: FizzBuzzList = 
    FizzBuzzList.create(values)
}
